crates/uv-redacted/src/lib.rs

```diff
@@ -1,21 +1,241 @@
+/// // Create a `DisplaySafeUrl` from a `&str`
+/// let mut url = DisplaySafeUrl::parse("https://user:password@example.com").unwrap();
+///
+/// // Display will mask secrets
+/// assert_eq!(url.to_string(), "https://user:****@example.com/");
+///
+/// // You can still access the username and password
+/// assert_eq!(url.username(), "user");
+/// assert_eq!(url.password(), Some("password"));
+///
+/// // And you can still update the username and password
+/// let _ = url.set_username("new_user");
+/// let _ = url.set_password(Some("new_password"));
+/// assert_eq!(url.username(), "new_user");
+/// assert_eq!(url.password(), Some("new_password"));
+///
+/// // It is also possible to remove the credentials entirely
+/// url.remove_credentials();
+/// assert_eq!(url.username(), "");
+/// assert_eq!(url.password(), None);
+/// ```
+#[derive(Clone, Eq, PartialEq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+#[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
+#[cfg_attr(feature = "schemars", schemars(transparent))]
+pub struct DisplaySafeUrl(Url);
+
+impl DisplaySafeUrl {
+    #[inline]
+    pub fn parse(input: &str) -> Result<Self, url::ParseError> {
+        Ok(Self(Url::parse(input)?))
+    }
+
+    /// Parse a string as an URL, with this URL as the base URL.
+    #[inline]
+    pub fn join(&self, input: &str) -> Result<Self, url::ParseError> {
+        self.0.join(input).map(DisplaySafeUrl::from)
+    }
+
+    /// Serialize with Serde using the internal representation of the `Url` struct.
+    #[inline]
+    pub fn serialize_internal<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        self.0.serialize_internal(serializer)
+    }
+
+    /// Serialize with Serde using the internal representation of the `Url` struct.
+    #[inline]
+    pub fn deserialize_internal<'de, D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        Url::deserialize_internal(deserializer).map(DisplaySafeUrl::from)
+    }
+
+    #[allow(clippy::result_unit_err)]
+    pub fn from_file_path<P: AsRef<std::path::Path>>(path: P) -> Result<DisplaySafeUrl, ()> {
+        Url::from_file_path(path).map(DisplaySafeUrl::from)
+    }
+
+    /// Remove the credentials from a URL, allowing the generic `git` username (without a password)
+    /// in SSH URLs, as in, `ssh://git@github.com/...`.
+    #[inline]
+    pub fn remove_credentials(&mut self) {
+        // For URLs that use the `git` convention (i.e., `ssh://git@github.com/...`), avoid dropping the
+        // username.
+        if self.0.scheme() == "ssh" && self.0.username() == "git" && self.0.password().is_none() {
+            return;
+        }
+        let _ = self.0.set_username("");
+        let _ = self.0.set_password(None);
+    }
+
+    /// Returns string representation without masking credentials.
+    #[inline]
+    pub fn to_string_with_credentials(&self) -> String {
+        self.0.to_string()
```

Comment:
Should this instead return a type that implements `Display`? It seems more flexible?